contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -1124,6 +1131,93 @@ validate_special_function(char *func_nsname, char *func_name, List* fargs, int n
 				}
 			}
 		}
+		else if (strlen(func_name) == 6 && strncmp(func_name, "concat", 6) == 0)
+		{
+			Oid		target_type_id = get_sys_varcharoid();
+			/* PG has limitation for max number of args = 100. */
+			if (nargs < 2 || nargs > 100)
+				ereport(ERROR,
+						(errcode(ERRCODE_UNDEFINED_FUNCTION),
+							errmsg("The concat function requires 2 to 100 arguments.")));
+			
```

Comment:
Suggestion: 
This implementation for supporting variadic functions(concat & concat_ws) looks messy, very specific, repetitive code and hard to maintain. We can extend our existing logic for special functions to support variadic special functions in a generic way, as follows
- Add a variable is_variadic in tsql_special_function struct to identify the variadic special function as follows
```
typedef struct tsql_special_function
{
	const char             *nsp;                              /* namespace of special function */
	const char             *funcname;                         /* name of special function */
	const char             *formatted_funcname;				  /* formatted name of special function */
	bool					is_variadic;
	int                     nargs;                            /* number of arguments of special function (for variadic function number of fixed arguments will be stored) */
	tsql_valid_arg_type_t   valid_arg_types[SFUNC_MAX_ARGS];  /* list for storing details of all the valid types supported for each arguments (for variadic function valid types of only fixed arguments will be stored) */
} tsql_special_function_t;
```
- For variadic function we can specify the fixed number of arguments and their details as follows
```
{"sys", "concat", "concat", true, 0, {}},
{"sys", "concat_ws", "concat_ws", true, 1, {{6, {"char","varchar","nchar","nvarchar","text","ntext"}, {InvalidOid, InvalidOid, InvalidOid, InvalidOid, InvalidOid, InvalidOid}}}}
```
- Instead of having different if blocks for almost same logic, a function can be created to validate variadic argument types, as follows. This function will be called only when special_func is variadic
```
static bool
validate_variadic_special_function_args(char *func_name, List* fargs, int nargs, int nfixed_args, Oid *input_typeids)
{
	Oid		input_type_id, base_type_id;
	Oid		sys_varcharoid = get_sys_varcharoid();

	if ((strlen(func_name) == 6 && strncmp(func_name, "concat", 6) == 0) ||
		 (strlen(func_name) == 9 && strncmp(func_name, "concat_ws", 9) == 0))
	{
		for (int i = nfixed_args; i < nargs; i++)
		{
			if (input_typeids[i] == UNKNOWNOID)
			{
				Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
				if (IsA(arg, Const) && ((Const *)arg)->constisnull)
					continue;
				else
					input_type_id = sys_varcharoid;
			}
			else
				input_type_id = input_typeids[i];

			base_type_id = get_immediate_base_type_of_UDT_internal(input_type_id);
			if (OidIsValid(base_type_id))
				input_type_id = base_type_id;

			/* Check if the input type is implicitly castable to varchar. */
			if (!can_coerce_type(1, &input_type_id, &sys_varcharoid, COERCION_IMPLICIT))
				ereport(ERROR,
					(errcode(ERRCODE_UNDEFINED_FUNCTION),
						errmsg("Implicit conversion from data type %s to varchar is not allowed.", format_type_be(input_typeids[i]))));
		}
		return true;
	}
	return false;
}
```

Reply:
Done